br/pkg/lightning/backend/external/onefile_writer.go

```diff
@@ -0,0 +1,240 @@
+	binary.BigEndian.AppendUint64(dataBuf[lengthBytes+encodedKeyLen:lengthBytes+encodedKeyLen], uint64(len(idxVal)))
+	copy(dataBuf[lengthBytes*2+encodedKeyLen:], idxVal)
+
+	w.kvLocations = append(w.kvLocations, kvLocation{
+		blockIdx: blockIdx,
+		offset:   off,
+		length:   int32(length)},
+	)
+	w.kvSize += int64(encodedKeyLen + len(idxVal))
+	w.batchSize += uint64(length)
+	return nil
+}
+
+// LockForWrite implements ingest.Writer.
+// Since flushKVs is thread-safe in external storage writer,
+// this is implemented as noop.
+func (w *OneFileWriter) LockForWrite() func() {
+	return func() {}
+}
+
+// Close closes the writer.
+func (w *OneFileWriter) Close(ctx context.Context) error {
+	if w.closed {
+		return errors.Errorf("writer %s has been closed", w.writerID)
+	}
+	w.closed = true
+	defer w.kvBuffer.destroy()
+	err := w.flushKVs(ctx, true)
+	if err != nil {
+		return err
+	}
+	w.logger.Info("close one file writer",
+		zap.String("writerID", w.writerID),
+		zap.Int("kv-cnt-cap", cap(w.kvLocations)),
+		zap.String("minKey", hex.EncodeToString(w.minKey)),
+		zap.String("maxKey", hex.EncodeToString(w.maxKey)))
+
+	w.kvLocations = nil
+
+	w.onClose(&WriterSummary{
+		WriterID:           w.writerID,
+		Seq:                0,
+		Min:                w.minKey,
+		Max:                w.maxKey,
+		TotalSize:          w.totalSize,
+		MultipleFilesStats: []MultipleFilesStat{w.multiFileStat},
+	})
+	return nil
+}
+
+func (w *OneFileWriter) recordMinMax(newMin, newMax tidbkv.Key, size uint64) {
+	if len(w.minKey) == 0 || newMin.Cmp(w.minKey) < 0 {
+		w.minKey = newMin.Clone()
+	}
+	if len(w.maxKey) == 0 || newMax.Cmp(w.maxKey) > 0 {
+		w.maxKey = newMax.Clone()
+	}
+	w.totalSize += size
+}
+
+func (w *OneFileWriter) flushKVs(ctx context.Context, fromClose bool) (err error) {
+	if len(w.kvLocations) == 0 {
+		return nil
+	}
+	
+	for _, pair := range w.kvLocations {
+		err = w.kvStore.addEncodedData(w.getEncodedKVData(pair))
+		if err != nil {
+			return err
+		}
+	}
+
+	w.kvStore.Close()
+	encodedStat := w.rc.encode()
+	_, err = w.statWriter.Write(ctx, encodedStat)
+	if err != nil {
+		return err
+	}
+
+	minKey, maxKey := w.getKeyByLoc(w.kvLocations[0]), w.getKeyByLoc(w.kvLocations[len(w.kvLocations)-1])
+	w.recordMinMax(minKey, maxKey, uint64(w.kvSize))
+
+	if fromClose {
+		w.multiFileStat.Filenames = append(w.multiFileStat.Filenames,
+			[2]string{w.dataFile, w.statFile},
+		)
+		w.multiFileStat.build(
+			[]tidbkv.Key{w.minKey},
+			[]tidbkv.Key{w.maxKey})
+
+		err1 := w.dataWriter.Close(ctx)
+		if err1 != nil {
+			w.logger.Error("Close data writer failed", zap.Error(err))
+			err = err1
+			return
+		}
+
+		err2 := w.statWriter.Close(ctx)
+		if err2 != nil {
+			w.logger.Error("Close stat writer failed", zap.Error(err))
+			err = err2
+			return
+		}
+	}
+	w.kvLocations = w.kvLocations[:0]
+	w.kvSize = 0
+	w.kvBuffer.reset()
+	w.rc.reset()
+	w.batchSize = 0
+	return nil
+}
+
+func (w *OneFileWriter) getEncodedKVData(pos kvLocation) []byte {
```

Comment:
The same as fileWriter, better to reuse the code.

Reply:
ok

Reply:
I will keep it, since it will affect the interface.